### ISSUE TITLE
Qwen3.5-9B and above trained under zero3 have most weights frozen

### ISSUE THREAD

**Issue by allen-dc** [OpenRLHF/OpenRLHF#1258]
`Actor.__init__` was calling [`set_z3_leaf_modules(self.model)`](https://github.com/OpenRLHF/OpenRLHF/blob/main/openrlhf/models/actor.py#L210), intended to work around a supposed DS-Z3-prefetch failure for Qwen3.5's hybrid decoder architecture (self_attn vs linear_attn per layer).

In practice DS's per-instance prefetch handles the hybrid layout correctly on its own, and the leaf marking silently drops gradients for ~390 of the ~417 inner params per decoder layer under Z3 for Qwen3.5-9B.

Only these tensors receive gradient updates when the leaf marking is active:
- lm_head.weight
- model.language_model.embed_tokens.weight
- model.language_model.norm.weight (final RMSNorm outside blocks)
- model.language_model.layers.X.linear_attn.norm.weight (small
  RMSNormGated inside GatedDeltaNet, only in linear_attn layers)

Everything else (mlp, self_attn, most linear_attn params, both layernorms inside each decoder block) stays bit-exact at initial weights across all training steps. This makes reported actor_grad_norm under Z3 systematically smaller than Z2 baseline.

How to reproduce:
- Train Qwen3.5-9B in zero3
- Compare the weights of the resulting checkpoint against the base model.

For this last step, we use the following utility script.

```
"""Standalone diagnostic — checks whether a Z3 checkpoint has the
"leaf module froze most params" bug.

Usage: only requires the initial base model weights and a Z3 checkpoint.
Does NOT need a Z2 baseline for comparison.

Bug fingerprint
---------------
Under Z3 with ``set_z3_leaf_modules(Qwen3_5DecoderLayer)`` and Qwen3.5,
gradients to most params INSIDE decoder layers are silently dropped.
Only a small subset of tensors ever receives a gradient:

  - lm_head.weight
  - model.language_model.embed_tokens.weight
  - model.language_model.norm.weight   (final RMSNorm, outside decoder blocks)
  - model.language_model.layers.X.linear_attn.norm.weight   (small 128-elem
    RMSNormGated inside GatedDeltaNet, only in linear_attn layers, escapes
    the leaf-module freeze for reasons unclear)

All other params inside decoder layers (mlp.*, self_attn.*, most
linear_attn.*, input_layernorm, post_attention_layernorm) stay at their
initial value BIT-EXACT across all training steps — no gradient ever
touched them.

What to look for
----------------
If the bug is present in the colleague's Z3 checkpoint you'll see:
  - ~27 tensors with non-zero delta from initial
  - ~390 tensors with EXACTLY zero delta from initial
  - The "frozen" tensor names are all the inner-decoder-layer params
    (mlp.*, self_attn.*, most linear_attn.*)
  - Even at global_step 100+, the frozen params are still exactly at init

If the bug is NOT present (fix applied or genuinely different setup):
  - All ~417 language-model tensors will have non-zero delta
  - Delta magnitudes will roughly correlate with lr × step_count × grad
    magnitude (larger for more-trained tensors)

Configuration
-------------
Set ``CHECKPOINT_PATH`` below to the colleague's Z3 checkpoint's
``model.safetensors``. Set ``MODEL_NAME`` if it's not Qwen3.5-9B.

Run:
    python check_frozen_params.py
"""

from __future__ import annotations

import glob
import os
import sys

import safetensors.torch
import torch
from huggingface_hub import snapshot_download


MODEL_NAME = os.environ.get("MODEL_NAME", "Qwen/Qwen3.5-9B")
CHECKPOINT_PATH = os.environ.get(
    "CHECKPOINT_PATH",
    "/path/to/colleague/z3_ckpt/global_stepN_hf/model.safetensors",
)


def _load_initial_state() -> dict[str, torch.Tensor]:
    """Read base model weights straight from safetensors — no model
    instantiation needed. Locates via huggingface_hub snapshot on disk;
    downloads if not cached."""
    snapshot_dir = snapshot_download(
        repo_id=MODEL_NAME, allow_patterns=["*.safetensors", "*.json"]
    )
    print(f"Initial weights from: {snapshot_dir}")
    shards = sorted(glob.glob(os.path.join(snapshot_dir, "*.safetensors")))
    state: dict[str, torch.Tensor] = {}
    for p in shards:
        state.update(safetensors.torch.load_file(p))
    print(f"  → loaded {len(state)} tensors from {len(shards)} shard(s)")
    return state


def main() -> None:
    if not os.path.exists(CHECKPOINT_PATH):
        print(f"ERROR: checkpoint not found: {CHECKPOINT_PATH}")
        print("Edit CHECKPOINT_PATH at the top of this script.")
        sys.exit(1)

    initial = _load_initial_state()

    print(f"Loading Z3 checkpoint: {CHECKPOINT_PATH}")
    z3 = safetensors.torch.load_file(CHECKPOINT_PATH)

    common = set(initial.keys()) & set(z3.keys())
    print(f"Common tensors: {len(common)} "
          f"(initial only: {len(initial) - len(common)}, "
          f"ckpt only: {len(z3) - len(common)})")

    moved: list[tuple[str, float]] = []
    exact_frozen: list[str] = []
    tiny_delta: list[tuple[str, float]] = []  # non-zero but suspiciously small

    for name in sorted(common):
        i = initial[name].float()
        z = z3[name].float()
        if i.shape != z.shape:
            continue

        # Delta relative to the tensor's typical value scale — a
        # frozen-then-copied param has delta EXACTLY zero (bit-exact).
        # A "properly updated" param has delta of order (lr × step_count).
        delta = (z - i).norm().item()
        base_norm = i.norm().item() + 1e-30
        rel_delta = delta / base_norm

        if delta == 0.0:
            exact_frozen.append(name)
        elif rel_delta < 1e-8:
            # Non-zero but many orders of magnitude below any plausible
            # training update — treat as effectively frozen.
            tiny_delta.append((name, rel_delta))
        else:
            moved.append((name, rel_delta))

    print()
    print(f"=== Result ===")
    print(f"  Moved (real training update):    {len(moved)}")
    print(f"  Exactly frozen (delta == 0):     {len(exact_frozen)}")
    print(f"  Tiny delta (< 1e-8 relative):    {len(tiny_delta)}")

    # The bug fingerprint check
    print(f"\n=== Bug-fingerprint check ===")
    frozen_all = set(exact_frozen) | {n for n, _ in tiny_delta}

    inner_frozen = [n for n in frozen_all if ".layers." in n]
    outer_frozen = [n for n in frozen_all if ".layers." not in n]
    inner_moved = [n for n, _ in moved if ".layers." in n]
    outer_moved = [n for n, _ in moved if ".layers." not in n]

    print(f"  Inner (per-layer) params: {len(inner_moved)} moved, {len(inner_frozen)} frozen")
    print(f"  Outer (embed/lm_head/final norm): {len(outer_moved)} moved, {len(outer_frozen)} frozen")

    # Specific fingerprint tensors we expect
    expected_moved_outer = {
        "lm_head.weight",
        "model.language_model.embed_tokens.weight",
        "model.language_model.norm.weight",
    }
    expected_bug_pattern = all(n in {n2 for n2, _ in moved} for n in expected_moved_outer)

    # Check the linear_attn.norm inner exception
    layer_norm_gated_moved = sum(
        1 for n, _ in moved if ".linear_attn.norm.weight" in n
    )

    print(f"  Outer 'always moves' tensors present in moved: "
          f"{expected_bug_pattern} (expected True)")
    print(f"  linear_attn.norm.weight tensors moved: {layer_norm_gated_moved} "
          f"(if bug: only these + outer; if healthy: many more inner tensors)")

    print()
    if len(inner_frozen) > 100 and len(outer_moved) >= 3 and layer_norm_gated_moved > 0:
        print("★ BUG FINGERPRINT MATCHES: most inner-decoder params frozen, "
              "outer + linear_attn.norm updating. This is the set_z3_leaf_modules "
              "gradient-freeze pattern.")
        print(f"  {len(inner_frozen)} inner tensors never received a gradient. "
              "The colleague's RL is training <5% of the model.")
    elif len(inner_frozen) < 20 and len(moved) > 400:
        print("✓ No freeze detected: ~all params moved, consistent with a "
              "healthy Z3 training run.")
    else:
        print("? Ambiguous result. Look at the numbers above:")
        print(f"  If most 'inner' tensors are frozen while 'outer' moved, "
              "it's the bug.")
        print(f"  If everything moved but by less than expected, could be "
              "a different issue.")

    # Print a few examples of each category for manual inspection
    print(f"\n--- First 10 frozen inner-layer tensors (if bug: entire decoder body) ---")
    for name in sorted(frozen_all)[:10]:
        if ".layers." in name:
            print(f"  {name}")

    print(f"\n--- First 10 moved tensors (if bug: only these types) ---")
    for name, rel in moved[:10]:
        print(f"  rel_delta={rel:.3e}  {name}")


if __name__ == "__main__":
    main()
```

Solution: Comment out the [`set_z3_leaf_modules(self.model)`](https://github.com/OpenRLHF/OpenRLHF/blob/main/openrlhf/models/actor.py#L210) line for Qwen3.5 family dense models. Did not test for MoE, however.

**Comment by Functionhx** [comment]
Found the root cause -- set_z3_leaf_modules registers full_backward_pre_hook on hybrid decoder layers, which steals gradient computation for ~390/417 inner params. Fixed by disabling only hybrid detection while keeping MoE leaf marking, applied to both actor and critic paths, with env var escape hatch. Have a PR ready.
